### ISSUE TITLE
Unable to update reader or preferences settings

### ISSUE THREAD

**Issue by Anzumatsuri** [Kareadita/Kavita#2937]
### Reader Settings

This issue occurs 100% of the time for me.

1. Open any manga or book
2. In reader, expand settings
3. Under Layout Mode, by default its set to Single
4. Set it to Double or Double (Manga)
5. Click Save Globally
6. Nothing happens. No toast notification on the bottom right notifying that it has been saved
7. Changing back to Single and save works fine.

Upon checking the dev tools, the `/api/users/update-preferences` returned `400 Bad Request.` The return response below:

```
{
    "type": "https://tools.ietf.org/html/rfc9110#section-15.5.1",
    "title": "One or more validation errors occurred.",
    "status": 400,
    "errors": {
        "preferencesDto": [
            "The preferencesDto field is required."
        ],
        "$.pageSplitOption": [
            "The JSON value could not be converted to API.Entities.Enums.PageSplitOption. Path: $.pageSplitOption | LineNumber: 0 | BytePositionInLine: 62."
        ]
    },
    "traceId": "00-c978ce553fad21f4132d767cc970c913-04a27573ba8b8052-00"
}
```
If its updated in User settings > Preference > Image Reader, it can be updated successfully.

### Preferences Settings

This is quite hard to reproduce. It does not occur 100% of the time. 

1. Open User Settings
2. Select the preference tab
3. Under Global Settings, even though my Locale was set to English, sometimes it would load as "Chinese (Simplified)". It won't appear most of the time, but it is reproducible if I spam refresh till it appears.

If It loads as "Chinese (Simplified)":
1. The Locale dropdown list cannot be expanded. 
2. Update any settings in preferences, click save.
3. Nothing happens. No toast notification on the bottom right notifying that it has been updated. 

Checking the dev tools again, the `/api/users/update-preferences` returned `400 Bad Request` as well. The return response below:

```
{
    "type": "https://tools.ietf.org/html/rfc9110#section-15.5.1",
    "title": "One or more validation errors occurred.",
    "status": 400,
    "errors": {
        "Locale": [
            "The Locale field is required."
        ]
    },
    "traceId": "00-cd52d0f54f6d63f0440033ad6ccb7cb3-8478fab545badeaa-00"
}
```

### What did you expect?

Both settings to be saved successfully.

### Kavita Version Number - If you don not see your version number listed, please update Kavita and see if your issue still persists.

0.8.1 - Stable

### What operating system is Kavita being hosted from?

Docker (Dockerhub Container)

### If the issue is being seen on Desktop, what OS are you running where you see the issue?

Windows

### If the issue is being seen in the UI, what browsers are you seeing the problem on?

Firefox, Chrome, Microsoft Edge

### If the issue is being seen on Mobile, what OS are you running where you see the issue?

iOS 17.4.1

### If the issue is being seen on the Mobile UI, what browsers are you seeing the problem on?

Safari, Chrome

### Additional Notes

- Running Kavita on Synology NAS DS1520+
- Kavita is front by Cloudflare, but issue is occurring on http://LAN_IP:33400 as well
- Been using since version 0.5.1.1 but I don't remember the issue coming up back then. I first noticed it on version 0.7.9.0.
- I have reported the preference settings issue on Discord previously when I was on version 0.7.10.2. A small check was added to the Locale in 0.7.13 but it did not fixed the issue.
- I have tried wiping the docker volumes and start from scratch, even for version 0.8.1. It did not do anything.

**Comment by DieselTech** [comment]
> If It loads as "Chinese (Simplified)":

We've seen this in the past, but it hasn't been reported by users in quite some time. If you run the `Bust Locale Cache` in the [admin panel tasks pages ](https://wiki.kavitareader.com/guides/admin-settings/tasks) can you see if that resets the Locale?

**Comment by Anzumatsuri** [comment]
> > If It loads as "Chinese (Simplified)":
> 
> We've seen this in the past, but it hasn't been reported by users in quite some time. If you run the `Bust Locale Cache` in the [admin panel tasks pages ](https://wiki.kavitareader.com/guides/admin-settings/tasks) can you see if that resets the Locale?

Hmm it doesn't prompt any toast notification whatsoever when I clicked the button, but upon checking the logs it seems like it indeed ran successfully:

![image](https://github.com/Kareadita/Kavita/assets/52699436/a87d0242-8869-4d07-9d61-b2fe45ea91b0)

But sadly the locale issue still occurs. ☹️ 
It appeared immediately when I navigate to the preference settings page right after. I refreshed once and it did not appear again. 
Though for the 2nd time it appears it took quite a number of refreshes. Pretty much literally spamming F5 till it changes to "Chinese (Simplified)."

**Comment by majora2007** [comment]
Can you try logging out of Kavita and back in?

**Comment by Anzumatsuri** [comment]
> Can you try logging out of Kavita and back in?

I ran `Bust Locale Cache` again and relog but still no luck. Locale still loads as `Chinese (Simplified)`.

**Comment by majora2007** [comment]
Sorry I should have clarified. What I'm trying to see is if you log out then back in again, if you are able to change the Locale and thus any other preferences, as that was your original issue. 

To my knowledge, this bug crops up from older installs upgrading, where something goes haywire and data on the client doesn't align with what's in the backend (aka locale is missing). I would also ask to clear your browser cache to ensure 100% an older version of UI code isn't playing a part here.

**Comment by Anzumatsuri** [comment]
Ah okay. My bad I should have explained it clearly as well. 

When I mentioned that it loads as `Chinese (Simplified)` it doesn't really means that it immediately loads as that. Most of the time the Locale loads correct in `English` but at times it just loads as `Chinese (Simplified)`. 

Spamming F5 to refresh was the only way for me to see if the issue still persist as on some refresh it'll randomly loads as `Chinese (Simplified)`.

Back to your suggestion, before logging out I went to the preference page for a peek and this time it immediately loads as `Chinese (Simplified)`. After relogging it loads correctly as `English`. Though spamming refresh still loads `Chinese (Simplified)` on some refreshes.

Locale issue aside, what about the Layout Mode issue?

**Comment by majora2007** [comment]
These 2 issues are related because your locale isn't being set correctly, meaning that when you go to save the preferences, it's not fully constructed. It might be something with your db, but to keep things simple, can you first fill out the user preferences by touching (selecting something else, then back to the original) every control on that page and hitting save? 

This should ensure everything gets set in the payload. Normally this shouldn't be an issue, but I bet you have some invalid data in one field and thus this issue is blocking you.

**Comment by Anzumatsuri** [comment]
Okay I have flipped every single dropdown list and switches 1 by 1 and save after each change. Nope its still broke on some refreshes.

**Comment by majora2007** [comment]
Wait when you saved, were you able to persist the settings like you reported with the layout mode?

**Comment by Anzumatsuri** [comment]
Yes the layout mode can be saved in user settings preference page. The error only occurs in the book/manga reader when I choose `Double` or `Double (Manga)`. 

![image](https://github.com/Kareadita/Kavita/assets/52699436/99c22798-b1b2-48ac-9d1e-3e7741f3c630)

Even if the default was set to one of the two in the user settings, the only setting that can be saved in the reader is setting it to `Single`.

Update:
Looks like the locale issue came up on mobile as well. Running iOS 17.4.1, it appeared on both Safari and Chrome.

**Comment by Anzumatsuri** [comment]
I've noticed the a difference between saving the layout mode in reader vs in preference page.
Below are the the request payload for saving the layout mode in both pages.

In Reader Settings:
```
{
  "readingDirection": 0,
  "scalingOption": 3,
  "pageSplitOption": null,              // This value will be null if "layoutMode" is set to 2 or 3
  "readerMode": 0,
  "layoutMode": 2,                        //  1 for Single, 2 for Double, 3 for Double (Manga)
  "emulateBook": false,
  "backgroundColor": "#000000",
  "swipeToPaginate": false,
  "autoCloseMenu": true,
  "showScreenHints": true,
  "bookReaderMargin": 15,
  "bookReaderLineSpacing": 100,
  "bookReaderFontSize": 100,
  "bookReaderFontFamily": "default",
  "bookReaderTapToPaginate": false,
  "bookReaderReadingDirection": 0,
  "bookReaderWritingStyle": 0,
  "theme": {
    "id": 1,
    "name": "Dark",
    "normalizedName": "dark",
    "fileName": "dark.scss",
    "isDefault": true,
    "provider": 1,
    "created": "2024-05-01T13:09:37.3565967",
    "lastModified": "2024-05-01T13:09:37.3563742",
    "createdUtc": "2024-05-01T05:09:37.3566615",
    "lastModifiedUtc": "2024-05-01T05:09:37.3564947"
  },
  "bookReaderThemeName": "Dark",
  "bookReaderLayoutMode": 0,
  "bookReaderImmersiveMode": false,
  "globalPageLayoutMode": 0,
  "blurUnreadSummaries": false,
  "promptForDownloadSize": true,
  "noTransitions": false,
  "collapseSeriesRelationships": true,
  "shareReviews": false,
  "locale": "en",
  "pdfTheme": 0,
  "pdfScrollMode": 0,
  "pdfLayoutMode": 0,
  "pdfSpreadMode": 0
}
```

In Preference Page:
```
{
  "readingDirection": 0,
  "scalingOption": 3,
  "pageSplitOption": 3,                  // This value isn't null, able to save all 3 layout modes
  "autoCloseMenu": true,
  "readerMode": 0,
  "layoutMode": 3,                        //  1 for Single, 2 for Double, 3 for Double (Manga)
  "showScreenHints": true,
  "backgroundColor": "#000000",
  "bookReaderFontFamily": "default",
  "bookReaderLineSpacing": 100,
  "bookReaderFontSize": 100,
  "bookReaderMargin": 15,
  "bookReaderTapToPaginate": false,
  "bookReaderReadingDirection": 0,
  "bookReaderWritingStyle": 0,
  "bookReaderLayoutMode": 0,
  "bookReaderThemeName": "Dark",
  "theme": {
    "id": 1,
    "name": "Dark",
    "normalizedName": "dark",
    "fileName": "dark.scss",
    "isDefault": true,
    "provider": 1,
    "created": "2024-05-01T13:09:37.3565967",
    "lastModified": "2024-05-01T13:09:37.3563742",
    "createdUtc": "2024-05-01T05:09:37.3566615",
    "lastModifiedUtc": "2024-05-01T05:09:37.3564947"
  },
  "bookReaderImmersiveMode": false,
  "globalPageLayoutMode": 0,
  "blurUnreadSummaries": false,
  "promptForDownloadSize": true,
  "noTransitions": false,
  "emulateBook": false,
  "swipeToPaginate": false,
  "collapseSeriesRelationships": true,
  "shareReviews": false,
  "locale": "en",
  "pdfTheme": 0,
  "pdfScrollMode": 0,
  "pdfSpreadMode": 0
}
```

**Comment by ghost** [comment]
If you first configure your library in English and then switch to Chinese after the initial installation, you will encounter this problem. This is my guess because I also encountered this problem.
@Anzumatsuri

**Comment by Anzumatsuri** [comment]
> If you first configure your library in English and then switch to Chinese after the initial installation, you will encounter this problem. This is my guess because I also encountered this problem. @Anzumatsuri

Nope, it was set to English from the start I did not change it even after a fresh reinstall. 

I went and switched to another language (French) and see if the issue still occurs after a few refresh, but so far it did not appear yet. I'll update again if it does.

I did noticed that some languages do not have any text values for the buttons, I'm not sure if its only occurring on my end.

![image](https://github.com/Kareadita/Kavita/assets/52699436/67219f8a-f586-4978-aa20-b5198861eb14)
![image](https://github.com/Kareadita/Kavita/assets/52699436/7640f344-35f7-40c3-9432-96d3c3b3ddce)

**Comment by Anzumatsuri** [comment]
Aww, sad to see that this cannot be reproduced on your end. Any other info you like to have/test that might help? I would love to see it get fixed.

Though from the looks of it, I may be the only one that's facing this issue. Well no rush, there are definitely other pressing issues that require your attention more.

I'm just wondering whether it could have been as simple as some permissions issue misconfigured or maybe its my Synology NAS/Container Manager just being dumb lol.

**Comment by majora2007** [comment]
It could be something screwy in your DB perhaps. You could show your `AppUserPreferences` table (or send me a dm with your DB) where we might be able to use some SQL to fix it. 

That's the only thing I can think of. I tried on my dev, prod and a new install and it was just working as expected. 

![image](https://github.com/Kareadita/Kavita/assets/735851/7d7c054e-ffb9-468d-9ebe-13ec0f090fc1)

**Comment by ghost** [comment]
I tested it and this error appeared many times.
However, I only tested changing the book background color and found that he performed well in English, which is often wrong in Chinese.
I didn't try comics because I didn't have them.
https://1drv.ms/v/s!AjUhMEyrls9_gbkH1F4a4y1UZaeXqA
@majora2007 
[kavita20240619.log](https://github.com/user-attachments/files/15893972/kavita20240619.log)

**Comment by ghost** [comment]
I tested the [demo site](https://demo.kavitareader.com) and have the same issue.
Here's my docker file
```
services:
  kavita:
    image: lscr.io/linuxserver/kavita:latest
    container_name: kavita
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Asia/Shanghai
    volumes:
      - /root/kavita/config:/config
      - /root/kavita/data:/data
    ports:
      - xxxx:5000
    restart: unless-stopped

**Comment by Anzumatsuri** [comment]
Here's my `AppUserPreferences` table:

| Id | AppUserId | AutoCloseMenu | BackgroundColor | BlurUnreadSummaries | BookReaderFontFamily | BookReaderFontSize | BookReaderImmersiveMode | BookReaderLayoutMode | BookReaderLineSpacing | BookReaderMargin | BookReaderReadingDirection | BookReaderTapToPaginate | BookReaderWritingStyle | BookThemeName | CollapseSeriesRelationships | EmulateBook | GlobalPageLayoutMode | LayoutMode | Locale | NoTransitions | PageSplitOption | PdfScrollMode | PdfSpreadMode | PdfTheme | PromptForDownloadSize | ReaderMode | ReadingDirection | ScalingOption | ShareReviews | ShowScreenHints | SwipeToPaginate | ThemeId |
| -- | --------- | ------------- | --------------- | ------------------- | -------------------- | ------------------ | ----------------------- | -------------------- | --------------------- | ---------------- | -------------------------- | ----------------------- | ---------------------- | ------------- | --------------------------- | ----------- | -------------------- | ---------- | ------ | ------------- | --------------- | ------------- | ------------- | -------- | --------------------- | ---------- | ---------------- | ------------- | ------------ | --------------- | --------------- | ------- |
| 1  | 1         | 1             | #000000         | 0                   | default              | 100                | 0                       | 0                    | 100                   | 15               | 0                          | 0                       | 0                      | Dark          | 1                           | 0           | 0                    | 1          | en     | 0             | 3               | 0             | 0             | 0        | 1                     | 0          | 0                | 3             | 0            | 1               | 0               | 1       |

**Comment by Anzumatsuri** [comment]
I know this is closed as you were not able to reproduce the issue.
Just wanted to inform that the issue has been fixed/resolved with the new UI revamp (Currently at 0.8.5.11).

Cheers!

**Comment by majora2007** [comment]
> I know this is closed as you were not able to reproduce the issue. Just wanted to inform that the issue has been fixed/resolved with the new UI revamp (Currently at 0.8.5.11).
> 
> Cheers!

This was closed as it's coming in the v0.8.6 release (which should release likely Monday, just waiting on localization contributors).
